### ISSUE TITLE
Comparison of Rivercast and DEQ Forcasts

### ISSUE THREAD

**Issue by emily-simmons07** [HARPgroup/baseflow_storage#170]
This [script](https://github.com/HARPgroup/baseflow_storage/blob/Emily_bf_workflow/deq_rivercast_comparison.R) compares Rivercast's ability to forecast with the DEQ's.  From the flow_csv, j_days are created with their corresponding percentiles. Then, the DEQ's 90-day forecasts are converted into percentiles and compared to the Rivercast percentiles, all against the obs_flow percentiles.

**Comment by emily-simmons07** [comment]
**Note:** 
- pct stands for percentile 
- The error is calculated from median_pct - obs_pct
- A negative error is due to an underprediction, and a positive error is due to an overprediction   
- Interpolation is happening to create the percentiles, which may limit the accuracy of the results 

**Table 1. 02059500:  Goose Creek near Huddleston**

| issue_date | forecast_week | forecast_date | deq_forecast (cfs)  |  obs_flow (cfs)  |  obs_pct  | rivercast_median_pct | deq_median_pct | rivercast_error             | deq_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |
| 2026-06-23 | 1 | 2026-06-30 | 30.22197  | 36.8 | 7.280000 | 3.063442 | 2.153444 | -4.2165578 | -5.1265565 |
| 2026-06-23  | 2 | 2026-07-07 | 27.38381  | 16.6  | 1.005791 | 6.206828 | 1.966486 | 5.2010370  | 0.9606955 |
| 2026-06-23   | 3   | 2026-07-14  | 24.81218 | 42.1 |13.102362 | 6.148007 | 2.383558   | -6.9543548  |  -10.7188042 |
| 2026-06-23 |   4  |  2026-07-21  | 22.48205 | 36.7 | 10.592334 | 5.897797 | 3.940729  | -4.6945374 | -6.6516056 | 
| 2026-06-23  |  5 |  2026-07-28  | 20.37075 | 30.2 | 4.188034  | 4.534430 | 1.210864  | 0.3463954 | -2.9771702  | 
| 2026-06-23   | 6  |  2026-08-04 | 18.45772  | 33.6  | 7.682540 | 4.359333| 2.217904 | -3.3232066 | -5.4646354 | 

<img width="623" height="308" alt="Image" src="https://github.com/user-attachments/assets/9755bd1b-0ea5-4ad8-8012-adc04d813b19" />

**Figure 1. Corresponding Hydrograph for Goose Creek** 

**Table 2. 02016000: Cowpasture River near Clifton Forge**

| issue_date | forecast_week | forecast_date | deq_forecast (cfs)  |  obs_flow (cfs)  |  obs_pct  | rivercast_median_pct | deq_median_pct | rivercast_error             | deq_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |
| 2026-06-23 | 1 | 2026-06-30 | 82.63425  | 107.0 | 19.000000 | 15.746037 | 1.908561 | -3.2539635| -17.091439 |
| 2026-06-23  | 2 | 2026-07-07 | 74.64996  | 88.6  | 11.000000 | 10.795565 | 4.324978 | -0.2044353  | -6.675022 |
| 2026-06-23   | 3   | 2026-07-14  | 68.40231 | 135.0 | 45.500000 | 14.586136 | 1.701156 | -30.9138641  |  -43.798844 |
| 2026-06-23 |   4  |  2026-07-21  | 63.44702 | 105.0 | 26.500000 | 15.623524 | 2.611754  | -10.8764763 | -23.888246 | 
| 2026-06-23  |  5 |  2026-07-28  | 59.47085 | 82.4 | 5.858586  | 9.324224 | 1.218964  | 3.4656386 | -4.639622  | 
| 2026-06-23   | 6  |  2026-08-04 | 56.24837  | 97.8  | 28.834639 | 11.448770 | 1.000000 | -17.3858696 | -27.834639 | 

<img width="614" height="313" alt="Image" src="https://github.com/user-attachments/assets/fd4d5acb-00ea-4c05-b65c-0f2f020dd20b" />

**Figure 2. Corresponding Hydrograph for Cowpasture**

**Initial thoughts:** Overall, both predictions are underpredicting more often than they are overpredicting. The Rivercast forecast seems to be closer than the deq forecast to the obs_pct, though I do not know how significant a role the interpolation is playing in these results.

**Comment by emily-simmons07** [comment]
**Adjusting the start date to see how effective the forecast is:**  

**Cowpasture**
Original start date: 06/23/2026
New start date: 06/28/2026 

The new start date is much closer to starting on a local trough, whereas the original start date was just chosen because that was right around when the Rivercast data started. I'd like to have the start date be 06/27/2026, but that csv date does not exist for the Rivercast data. 

<img width="448" height="323" alt="Image" src="https://github.com/user-attachments/assets/5326c964-7676-4133-867c-3b92bdd1205e" />

**Figure 1: Start Date 06/28/26** 

**Table 1:** 
| issue_date | forecast_week | forecast_date | deq_forecast (cfs)  |  obs_flow (cfs)  |  obs_pct  | rivercast_median_pct | deq_median_pct | rivercast_error             | deq_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |
| 2026-06-28 | 1 | 2026-07-05 | 90.21644  | 93.3 | 13.00000 | 8.389150 | 12.065588 | -4.6108503 | -0.9344123 |
| 2026-06-28  | 2 | 2026-07-12 | 80.50332  | 170.0  | 61.50000 | 4.290502 | 4.750996 | -57.2094984  | -56.7490043 |
| 2026-06-28   | 3   | 2026-07-19  | 72.99132 | 84.6 | 8.00000 | 8.776830 | 3.799132 | 0.7768297  | -4.2008684 |
| 2026-06-28 |   4  |  2026-07-26  | 67.09274 | 82.6 | 7.20000 | 9.269992 | 2.578049  | 2.0699919 | -4.6219505 | 
| 2026-06-28  |  5 |  2026-08-02  | 62.40032 | 79.5 | 7.55477  | 5.828240 | 1.000000  | -1.7265299 | -6.5547703 | 

By changing the start date, both the Rivercast and deq errors decreased overall. The deq forecast has an extremely accurate 7-day forecast. The only issue with this start date is that on week two (07-12-2026), there is a significant spike in observed flow to 107 cfs from 89.9 cfs on 07-09-2026, which neither forecast can account for. 

**Goose Creek** 

Original start date: 06/23/2026
New start date: 07/07/2026

<img width="618" height="292" alt="Image" src="https://github.com/user-attachments/assets/5b25b5b5-c51a-44c9-8c7c-1cc9d871e346" />

**Figure 2: Start Date: 07/07/26**

**Table 2:** 

| issue_date | forecast_week | forecast_date | deq_forecast (cfs)  |  obs_flow (cfs)  |  obs_pct  | rivercast_median_pct | deq_median_pct | rivercast_error             | deq_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |  -------- |
| 2026-07-07 | 1 | 2026-07-14 | 15.165590 | 42.1 | 13.102362 | 3.262577 | 1 | -9.839785 | -12.102362 |
| 2026-07-07  | 2 | 2026-07-21 | 13.741380  | 36.7  | 10.592334 | 4.976262 | 1 | -5.616073  | -9.592334 |
| 2026-07-07   | 3   | 2026-07-28  | 12.450919 | 30.2 | 4.188034 | 5.431576 | 1 | 1.243542  | -3.188034 |
| 2026-07-07 |  4 |  2026-08-04 | 11.281646 | 33.6 | 7.682540 | 2.206313 | 1   | -5.476227 | -6.682540 | 

I would have liked to have the start date on 06/28/26, but there are no flow records for the 28th or the 29th, and the 30th is a peak, not a trough. The next trough was 07/07/26. The only issue with this start date is that all the deq forecasted flows fall within the lowest historic percentile within the deq_median_pct. Overall, this start date for Goose Creek decreased the accuracy of both projections compared with the original start date of 06/23/26. 

In progress!

**Comment by emily-simmons07** [comment]
<img width="695" height="312" alt="Image" src="https://github.com/user-attachments/assets/de56b624-b6a3-4409-8f92-308e6e4fb775" />

 **02059500: Goose Creek near Huddleston**
Start Date: 07/07/26
End Date: 8/25/26

**Table 1: Week One Error**

| issue_date | forecast_week | forecast_date | obs_pct  | rivercast_median_pct | rivercast_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  
| 2026-07-07 | 1 | 2026-07-14 | 13.102362  | 3.262577 | -9.839785 | 
| 2026-07-07   | 2 | 2026-07-21 | 10.689655  | 4.976262  | -5.713394 | 
| 2026-07-07    | 3   | 2026-07-28  | 5.698925 | 5.431576 | -0.267349 | 
| 2026-07-07  |   4  |  2026-08-04  | 13.091549 | 2.206313 | -10.885236 | 
| 2026-07-07   |  5 |  2026-08-11 | 7.756757 | 6.332922 | -1.423835  | 
| 2026-07-07 | 6 | 2026-08-18 | 5.966184  | 3.989808   | -1.976376  |
| 2026-07-07 | 7 | 2026-08-25 |  4.044321  | 2.372997  |  -1.671324 |

**Table 2: Week Two Error** 

| issue_date | forecast_week | forecast_date | obs_pct  | rivercast_median_pct | rivercast_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  
| 2026-07-14 | 1 | 2026-07-21 | 10.689655    | 2.20077753 | -8.488878 | 
| 2026-07-14 | 2 | 2026-07-28 | 5.698925  | 4.50508690 | -1.193838 | 
| 2026-07-14 | 3   | 2026-08-04  | 13.091549  | 7.29023838 | -5.801311 | 
| 2026-07-14 |   4  |  2026-08-11  | 7.756757  | 5.60204744 | -2.154709 | 
| 2026-07-14  |  5 |  2026-08-18 | 5.966184 | 7.12161064 | 1.155427 | 
| 2026-07-14 | 6 | 2026-08-25 | 4.044321 | 5.96639872 | 1.922077 |

**Table 3: Week Three Error** 

| issue_date | forecast_week | forecast_date | obs_pct  | rivercast_median_pct | rivercast_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  
| 2026-07-21| 1 | 2026-07-28 | 5.698925  | 4.561699 | 1.638259 | 
| 2026-07-21 | 2   | 2026-08-04  | 13.091549  | 7.640503 | -7.093460 | 
| 2026-07-21 |  3  |  2026-08-11  | 7.756757  | 9.038583 | -1.059464 | 
| 2026-07-21  | 4 |  2026-08-18 | 5.966184 | 8.850924 | 1.654472 | 
| 2026-07-21 | 5 | 2026-08-25 | 4.044321 | 8.132664 | 1.050074 |

**Table 4: Week Four Error** 

| issue_date | forecast_week | forecast_date | obs_pct  | rivercast_median_pct | rivercast_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  
| 2026-07-28 | 1   | 2026-08-04  | 13.091549  | 10.441963 | -2.6495861 | 
| 2026-07-28 |  2 |  2026-08-11  | 7.756757  | 7.986245 | 0.2294884 | 
| 2026-07-28  | 3 |  2026-08-18 | 5.966184 | 10.430584 | 4.4644004 | 
| 2026-07-28 | 4 | 2026-08-25 | 4.044321 | 9.642216 | 5.5978944 |

**Table 5: Week Five Error** 

| issue_date | forecast_week | forecast_date | obs_pct  | rivercast_median_pct | rivercast_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  
| 2026-08-04 |  1 |  2026-08-11  | 7.756757  | 4.493699 | -3.263058 | 
| 2026-08-04 | 2 |  2026-08-18 | 5.966184 | 4.288189 | -1.677995 | 
| 2026-08-04 | 3 | 2026-08-25 | 4.044321 | 5.795599 | 1.751278 |

**Table 6: Week Six Error** 

| issue_date | forecast_week | forecast_date | obs_pct  | rivercast_median_pct | rivercast_error |
| -------- | -------- | -------- |  -------- |  -------- |  -------- |  
| 2026-08-11| 1 |  2026-08-18 | 5.966184 | 10.152960 | 0.1226026 | 
| 2026-08-11| 2 | 2026-08-25 | 4.044321 | 5.901470 | 1.2833712 |

This analysis aims to collect data on how the Rivercast predictions are performing. A manually selected start date of July 7th was chosen, and then the forecast was run on the 7th,14th, 21st, 28th, 4th, and 11th in an attempt to see which weeks are performing better than others and if there were any clear patterns. 

Looking at an end date of 2026-08-18, for example, in the first forecast the error is -1.976376, then 1.155427, then 1.654472, then 4.4644004, and finally -1.677995. The forecast is the least accurate, with a start date of 2026-07-28 (week four), likely due to precipitation. Overall, it is difficult to extract any overarching patterns from this analysis without further testing.

**Comment by emily-simmons07** [comment]
This [Modeling Streamflow Drought](https://water.usgs.gov/vizlab/modeling-drought/) outlines the USGS processes for creating the River Drought Cast. 

**USGS percentile classifications:** 

- Moderate streamflow drought (10th - 20th percentile, USDM D1 drought)
- Severe streamflow drought (5th - 10th percentile, USDM D2 drought)
- Extreme streamflow drought (< 5th percentile, USDM D3 drought)

**How the USGS Generates Predictions**

_Median Prediction:_ The model uses a symmetric loss function which equally penalizes over and under predictions at the same rate. 

_95% quantile Prediction:_ The model uses an asymmetric loss function that penalizes under predictions more heavily than over predictions. This directional penalty results in a plot where the left slope (underestimate) is steeper than the right (overestimate). 

_5% quantile Prediction:_ The model uses an asymmetric loss function that penalizes overpredictions more than under predictions. This produces a loss function with a steeper right slope.

**Limitations** 

- The training data set is restricted to the observed streamflow values below the 50th percentile in order to improve low streamflow percentiles. The model cannot capture uncertainty above the 50th percentile  
- Generally, less accurate the further out the forecast

**Comment by emily-simmons07** [comment]
[Machine learning-generated streamflow drought forecasts for the conterminous United States (CONUS): developing and evaluating an operational tool to enhance sub-seasonal to seasonal streamflow drought early warning for gaged locations](https://www.frontiersin.org/journals/water/articles/10.3389/frwa.2025.1709138/full)

**Inputs into the machine learning (ML):** 
1. Daily streamflow data from 1981 to 2020 from the USGS. Then this daily flow data was turned into a 7-day average daily flow and then converted into a percentile. They created the percentiles using only values from a 30-day window surrounding that day, across all years of record. By doing this, they hope to account for seasonally relevant data.  
2. For land surface model output on soil moisture, they used NLDAS2
3. For gridded meteorological variables, they used gridMET
4. For snow water equivalent, they used model data from the National Snow and Ice Data Center Distributed Active Archive Center
5.  For forecast meteorology, they used the Global Ensemble Forecast System for days 1-10, and for 1-3 months they used the North American Multimodel Ensemble
6. Reservoir inflow, storage, and release were obtained for more than 500 sites with long-term records from ResOpsUS

Once weekly values were calculated, a number of rolling means (ranging from 30 to 365 days) and percentile transformations were performed. 

**Overall Method** 
In progress!
